### PR TITLE
refactors statpanel to use tgui API

### PR DESCRIPTION
**Pull request by azrael-x86**
## About The Pull Request

refactors the status panel to utilize the [tgui/byond communication APIs](https://github.com/tgstation/tgstation/blob/master/tgui/docs/tgui-for-custom-html-popups.md) instead of passing along href data, as well as converts the entirety of it into a `datum/tgui_window`

## Why It's Good For The Game

communication between DM and html has never been pretty: utilizing `location.href` to provide a maximum of 2048 characters of string typed data. the stat panel serves as a suitable test subject for this relatively new API

### CODE REVIEW

**Comment by JohnFulpWillard** [comment]
https://github.com/tgstation/tgstation/blob/master/.github/guides/STANDARDS.md

image

Make sure it is still possible to use HTML

**Comment by magatsuchi** [comment]
> https://github.com/tgstation/tgstation/blob/master/.github/guides/STANDARDS.md
> 
> image
> 
> Make sure it is still possible to use HTML

mothblocks concurs that unless MSO has a problem with it, utilizing `datum/tgui_window` in place of normal HTML writing (aka using the output operator) is fine

**Comment by magatsuchi** [comment]
(speaking of, @MrStonedOne you should look at this)

**Comment by stylemistake** [comment]
I'm turning it into a draft since it needs some polish.

TGUI rule does not apply here since this is a single page HTML with no assets.

File: tgui/public/tgui.html [hunk 1]
```diff
@@ -9,6 +9,7 @@
 
 <!-- Early setup -->
 <script type="text/javascript">
+var Byond = window.Byond = {};
```
Comment (stylemistake) [inline tgui/public/tgui.html:12]: Why did you move this?

**Comment by stylemistake** [comment]
Things I'd like you to address:

Instead of using one big Byond.subscribe and a bunch of separate functions, you should turn the callback function into this in-place construct, which makes it clear which functions are the actual functions, and which functions are callbacks to DM messages.

```js
Byond.subscribeTo(type, function (payload) {
  // do not call an external function here
  // but instead handle all logic in place
});
```

e.g.

```js
function update_sdql2(S) {
	sdql2 = JSON.parse(S);
	if(sdql2.length > 0 && !verb_tabs.includes("SDQL2")) {
		verb_tabs.push("SDQL2");
		addPermanentTab("SDQL2");
	}
	if(current_tab == "SDQL2")
		draw_sdql2();
}
```

This would be better as:
```js
Byond.subscribeTo('update_sdql2', function (sdql2) {
	if(sdql2.length > 0 && !verb_tabs.includes("SDQL2")) {
		verb_tabs.push("SDQL2");
		addPermanentTab("SDQL2");
	}
	if(current_tab == "SDQL2")
		draw_sdql2();
});
```

Notice that you don't need to JSON.parse anything, nor do you need to manually json_encode it in DM, this is because objects that are sent via tgui message protocol are already json encoded, so you're basically json-encoding things twice, and there's no need to do that. What you send as payload is what you get in subscribe.

So you need to rework both subscribe functions and json_encode things such as this:

```dm
target.stat_panel.send_message("update_tickets", list(tickets = json_encode(ahelp_tickets)))
```

---

You need to move JS to a separate file because it's a lot of code, it messes up git diffs, it will be more convenient that way. Currently I see that `inline_js` doesn't work for you, I'll try to see what's the issue, ping me on Discord so that we could figure it out.

---

You should replace `[addEventListenerKey]` with `.addEventListener` and `[textContentKey]` with `.textContent` in all because polyfills for this API are included, this is the big reason why we're doing this.

---

```js
window.onload = function () {
	setupSubscribe();
	NotifyByondOnload();
};

function NotifyByondOnload() {
	Byond.command("Panel-Ready")
}
```

Byond.subscribe() and subscribeTo() are not dependent on the `load` event, so they can subscribe right away, without being contained within onload callback or a wrapper function.

`Panel-Ready` verb is not doing anything useful at all:

```dm
/client/verb/panel_ready()
	set name = "Panel Ready"
	set hidden = TRUE

	statbrowser_ready = TRUE
	init_verbs()
```

Ready check is already implemented as part of tgui and you already used `stat_panel.is_ready()` in many places, so you need to only call `Update-Verbs`. And NotifyByondOnload wrapper function is unnecessary at all, just send the command directly in onload callback.

---

Many statpanel verbs are used as remote functions, they're pretty much all hidden, so instead you can use `Byond.sendMessage` and handle it in your choice of proc, which you subscribe to messages via `window.subscribe(src, .proc/on_stat_panel_message)`, smth like that.

**Comment by MrStonedOne** [comment]
> > https://github.com/tgstation/tgstation/blob/master/.github/guides/STANDARDS.md
> > image
> > Make sure it is still possible to use HTML
> 
> mothblocks concurs that unless MSO has a problem with it, utilizing `datum/tgui_window` in place of normal HTML writing (aka using the output operator) is fine

na, this actually isn't fine.

the chat box being tgui only makes it past my requirement because when tgui fails to load the chat box falls back to a byond output control.

because of how we did browser stat panels, this can not be done for them.

If tgui_panel didn't have the bluescreen stuff, or typescript, or any of the other cancerous "halt and break everything on 1 tiny error" it could be workable for this usecase, but right now this is only gonna increase unreliability when before, on a minor script error, it would still try to keep chugging along.

**Comment by stylemistake** [comment]
> If tgui_panel didn't have the bluescreen stuff

We can turn the bluescreen feature off for all panels as part of this PR, so non-issue.

> tgui fails to load

This particular "tgui" doesn't have assets which is the primary reason why UIs can fail to load, so stat panel can't possibly "fail to load".

**Comment by MrStonedOne** [comment]
The primary reason uis fail to load is because they timeout loading because byond on some computers will randomly take 60 seconds to create the `browser` element while a white screen (from the empty `window` skin element with no contents) laughs at your face.

Luckily it seems the html panel system used by tgui chat doesn't have the annoying (but sadly necessary) timeout system, but given that the pieces this interfaces uses exist in mostly plain javascript in the interface, i do want to ensure that react failing to init for *any* reason doesn't inappropriately block the interface from working.

**Comment by stylemistake** [comment]
> because byond on some computers will randomly take 60 seconds to create the browser element while a white screen (from the empty window skin element with no contents) laughs at your face.

That's your assumption, because general tgui windows will appear as white while they're fetching assets, and this means they might be actually running scripts already, you just don't see it. (I wanted to make them show a spinner at some point to distinguish it from just a white screen, but anyway).

I can guarantee with 99% probability it's caused by assets being slow or failing to download, which causes a retry (which is also invisible, until a certain point, then it fails with a blue screen).

File: tgui/public/tgui.html (hunk 1 above)
Comment (stylemistake) [reply tgui/public/tgui.html:12]: I moved it back where it was.

**Comment by MrStonedOne** [comment]
> That's your assumption, because general tgui windows will appear as white while they're fetching assets, and this means they might be actually running scripts already, you just don't see it. (I wanted to make them show a spinner at some point to distinguish it from just a white screen, but anyway).
> 
> I can guarantee with 99% probability it's caused by assets being slow or failing to download, which causes a retry (which is also invisible, until a certain point, then it fails with a blue screen).


No, its fucking not an assumption. I literally monitored the process with winspy, the mshtml browser hwnd control doesn't get created until the end of the delay,

I understand this bug better then you, because I *get* this fucking bug, and have for 5 years.

**Comment by stylemistake** [comment]
:ok:

**Review by MrsTonedOne — APPROVED** [review]

File: html/statbrowser.js [hunk 2]
```diff
@@ -0,0 +1,1066 @@
...
+	}
+});
+
+Byond.subscribeTo('remove_listedturf', remove_listedturf);
+
+Byond.subscribeTo('remove_sdql2', remove_sdql2);
+
+Byond.subscribeTo('remove_mc', remove_mc);
+
+Byond.subscribeTo('add_verb_list', add_verb_list(payload));
```
Comment (stylemistake) [inline html/statbrowser.js:1066]:
```suggestion
Byond.subscribeTo('add_verb_list', add_verb_list);
```

File: html/statbrowser.js [hunk 3]
```diff
@@ -0,0 +1,1066 @@
...
+		draw_tickets();
+		draw_interviews();
+	} else if (tab == "SDQL2") {
+		draw_sdql2();
+	} else if (tab == turfname) {
+		draw_listedturf();
+	} else {
+		statcontentdiv[textContentKey] = "Loading...";
+	}
+	window.location.href = "byond://winset?statbrowser.is-visible=true";
```
Comment (stylemistake) [inline html/statbrowser.js:347]: Can be refactored to Byond.winset

File: html/statbrowser.js [hunk 4]
```diff
@@ -0,0 +1,1066 @@
...
+		}
+		var __nativeSI__ = window.setInterval;
+		window.setInterval = function (vCallback, nDelay /*, argumentToPass1, argumentToPass2, etc. */) {
+			var aArgs = Array.prototype.slice.call(arguments, 2);
+			return __nativeSI__(vCallback instanceof Function ? function () {
+				vCallback.apply(null, aArgs);
+			} : vCallback, nDelay);
+		};
+	}, 0, 'test');
+}())
```
Comment (stylemistake) [inline html/statbrowser.js:61]: This entire polyfill appears to be unused

**Comment by stylemistake** [comment]
tgchat will continue to have bluescreens, since that's how it's been for a long time, and fix chat verb exists for fixing it

File: code/controllers/subsystem/statpanel.dm [hunk 5]
```diff
@@ -258,6 +254,20 @@ SUBSYSTEM_DEF(statpanels)
...
+/**
+ * Handles incoming messages from the stat-panel TGUI.
+ */
+/client/proc/on_stat_panel_message(type, payload)
+	if(type == "Update-Verbs")
+		init_verbs()
+		return
+	if(type == "Remove-Tabs")
+		panel_tabs -= payload["tabs"]
+		return
```
Comment (stylemistake) [inline code/controllers/subsystem/statpanel.dm:269]: Currently unused, iirc the reason was:

> ahh
> ill probably need to make the queue into an associative list
> so i can make it use Byond.sendMessage instead of Byond.command
> to remove those verbs

From what I gathered, the `connected` flag is only false when not fully initialized, then it's set to true, and it never resets to false, so I feel like this whole command queuing construct is built for nothing.

``` js
// Any BYOND verb call must go through this, as if a verb is sent during reconnect then
// it will cause the reconnect to fail.
// This function will either call immediately, or queue until
// BYOND confirms we are connected.
function send_byond_command(command) {
```

So maybe this queuing function doesn't have to exist at all, which will free you up to use Byond.command() and Byond.sendMessage() directly.

**Comment by stylemistake** [comment]
^ doesn't help with loading unloadable icons, and causes new bugs

**Comment by san7890** [comment]
During Round 183219, I literally could not click any of the tabs to get the following changed (in order to see the contents of that tab).

image

I had died and was revived at this point, so maybe something in relation to having the tabs+verbs available to me?

**Comment by magatsuchi** [comment]
> During Round 183219, I literally could not click any of the tabs to get the following changed (in order to see the contents of that tab).
> 
> image
> 
> I had died and was revived at this point, so maybe something in relation to having the tabs+verbs available to me? Sometimes as I moved around, the contents of that above screenshot would flash away and actually show me what I wanted to see (the "Status" Panel), but only for a second at most.

did the tgui log show anything?
